code/modules/spells/targeted/street_alchemy.dm

```diff
@@ -0,0 +1,140 @@
+	level_max = list(Sp_TOTAL = 4, Sp_SPEED = 3, Sp_POWER = 1, Sp_RANGE = 1)
+
+/spell/targeted/alchemy/apply_upgrade(upgrade_type)
+	switch(upgrade_type)
+		if(Sp_SPEED)
+			return quicken_spell()
+		if(Sp_POWER)
+			spell_levels[Sp_POWER]++
+			name = "Permeating " + name
+			return "Thrown elixirs now transfer reagents."
+		if(Sp_RANGE)
+			spell_levels[Sp_RANGE]++
+			return "You now pilfer in an area around the target."
+
+/spell/targeted/alchemy/get_upgrade_info(upgrade_type)
+	switch(upgrade_type)
+		if(Sp_SPEED)
+			return "Reduce this spell's cooldown."
+		if(Sp_POWER)
+			return "Thrown elixirs transfer their contents into living targets."
+		if(Sp_RANGE)
+			return "Pilfers reagents in an area around the target."
+
+/spell/targeted/alchemy/cast(list/targets, mob/user)
+	for(var/target in targets)
+		if(spell_levels[Sp_RANGE])
+			aoeAlchemy(target, user)
+			return
+		singleAlchemy(target, user)
+
+/spell/targeted/alchemy/proc/singleAlchemy(target, mob/user)
+	var/obj/item/weapon/reagent_containers/pill/streetAlchElixir/A = null
+	if(spell_levels[Sp_POWER])
+		A = new /obj/item/weapon/reagent_containers/pill/streetAlchElixir/hypoElixir(src)
+	else
+		A = new /obj/item/weapon/reagent_containers/pill/streetAlchElixir(src)
+	if(user.find_empty_hand_index())
+		user.put_in_hands(A)
+	if(iscarbon(target))
+		mobAlchemy(target, user, A)
+	if((istype(target, /obj/structure)) || (istype(target, /obj/item/weapon/storage)) || (istype(target, /obj/item/weapon/reagent_containers)) || (istype(target, /obj/machinery)))
+		itemAlchemy(target, user, A)
+	if(A.is_empty())
+		to_chat(user, "You fail to perform alchemy.")
+		user.drop_item(A, force_drop = 1)
+		qdel(A)
+		return
+	to_chat(user, "You perform alchemy.")
+	playsound(user, "sound/effects/bubbles.ogg", 75, 1)
+
+/spell/targeted/alchemy/proc/aoeAlchemy(target, mob/user)
+	var/obj/item/weapon/reagent_containers/pill/streetAlchElixir/A = null
+	if(spell_levels[Sp_POWER])
+		A = new /obj/item/weapon/reagent_containers/pill/streetAlchElixir/hypoElixir(src)
+	else
+		A = new /obj/item/weapon/reagent_containers/pill/streetAlchElixir(src)
+	if(user.find_empty_hand_index())
+		user.put_in_hands(A)
+	for(var/T in range(2, target))
+		if(iscarbon(T))
+			mobAlchemy(T, user, A)
+		if((istype(T, /obj/structure)) || (istype(T, /obj/item/weapon/storage)) || (istype(T, /obj/item/weapon/reagent_containers)) || (istype(T, /obj/machinery)))
+			itemAlchemy(T, user, A)
+	if(A.is_empty())
+		to_chat(user, "You fail to perform alchemy.")
+		user.drop_item(A, force_drop = 1)
+		qdel(A)
+		return
+	to_chat(user, "You perform alchemy.")
+	playsound(user, "sound/effects/bubbles.ogg", 75, 1)
+
+/spell/targeted/alchemy/proc/mobAlchemy(var/mob/living/carbon/C, mob/user, var/A)
+	for(var/S in get_contents_in_object(C))
+		if(istype(S, /obj/item/weapon/reagent_containers))
+			var/obj/item/weapon/reagent_containers/F = S
+			F.reagents.trans_to(A, F.reagents.total_volume)
```

Comment:
all of these various loops should break if `reagents.is_full()` to stop the loop from continuing after your elixir is filled; I'd also recommend making A a more descriptive name